Flexecutioners: user-composable execution control

rwcarlsen opened the pull request:
ref #5229 and #18180.

## Reason
<!--Why do you need this feature or what is the enhancement?-->

## Design
<!--A concise description (design) of the enhancement.-->

## Impact
<!--Will the enhancement change existing APIs or add something new?-->


<!--
If this PR implements a bug fix, please create an issue for the bug and reference the issue.
-->

rwcarlsen committed:
add new runner system

ref #5229, #18180

rwcarlsen committed:
trying to fix up secant solve to work with refactored multiapp solve config params

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/834650/) on 164b90e wanted to post the following:

A change of the following file(s) triggered this check:

  libmesh

The following file(s) are unchanged:

  conda/mpich/conda_build_config.yaml
  conda/libmesh/meta.yaml

The libmesh submodule was changed but the conda build config was not

YaqiWang commented:
Looks overall pretty good. Deriving runners from `Executioner` does not seem right. What I'd like to see in this PR is to implement `FEProblemSolve` into a runner and create a test actually use it. We have lots of tests using `Steady` not doing Picard and mesh adaptation, that can be used to test this runner. The current execution_on of `initial` and `final` could cause some headache. Essentially this system needs to be in a decent stable shape so that we can implement complicated runs like depletion on top. Will talk more during our discussion.

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/834820/) on 739d3e0 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/18966/site)

This comment will be updated on new commits.

GiudGiud commented:
Should multiapps & mesh generators be 'runners' as well ?

rwcarlsen committed:
tweak result printing

rwcarlsen committed:
add runner unit testing

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/838325/) on 66e93a5 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/18966/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/18966/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format a5061ff5947f8418504977a4c28c154cc1932fea`

rwcarlsen committed:
delete test for feature I deleted

rwcarlsen committed:
remove erroneous test

Guillaume investigated this test and its results and confirmed that its
setup and answers files (including curreng gold files) are
invalid+incorrect both prior to this PR and after this PR.

ref #18990

rwcarlsen committed:
add doc comments to runner class

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/840321/) on 02032e2 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/18966/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/18966/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format ce01b12ebb1794f0cd12c017e75d95d23a22d878`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/840353/) on 35375bf wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/18966/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/18966/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format ce01b12ebb1794f0cd12c017e75d95d23a22d878`

rwcarlsen commented:
@idaholab/moose-team - soliciting opinions on how to name the future executioner replacement - eventually I'd like to have this be the executioner system.  But while we get things sorted out with this "experimental" feature - it will be a separate system/syntax.  Currently, I have the new class/system named "Runner".  @friedmud suggested perhaps "Flexecutioner".  I'm open to ideas.

friedmud commented on framework/doc/content/source/runners/Runner.md:
````diff
@@ -0,0 +1,184 @@
+A hypothetical example is this:
+
+```
+[Runner]
````
I would want this to be plural: `Runners`

friedmud commented on framework/doc/content/source/runners/Runner.md:
```diff
@@ -0,0 +1,184 @@
+
+This runner would simply generate the init+solve+refine trio of runners
+programmatically - hiding the runner structure from the user.  By default the
+last runner listed in the `Runner` block becomes the master/primary runner.
```
This shouldn't be necessary - we should find the root runner by sorting the graph using the dependency resolver.

rwcarlsen replied:
I agree.  The trick is that I'd like runners to be able to grab their child/sub runners in their constructors - but that only works if we create the objects themselves in the correct order.  But we can't run any normal dependency resolver interface or similar techniques on objects that haven't been created yet.  So I'm not sure the best approach.  Have we resolved this in e.g. mesh generators in any way that doesn't require inter-object binding to go outside of the constructor?

roystgnr commented:
How about "Executor"?  Same "one who executes" denotation as "executioner", but with more "carries out instructions" connotations and fewer "kills people" connotations.

friedmud commented on framework/include/executioners/Runner.h:
```diff
@@ -0,0 +1,165 @@
+  /// runner.  Custom runner behavior should be localized to this function.  If
+  /// you are writing a runner - this is basically where you should put all your
+  /// code.
+  virtual Result gogogadget() = 0;
```
Can we get a different name here?  Even for experimental code... it would be better to have something else.

friedmud replied:
`runImplementation`, `actuallyRun`, `sprint`, `start`, `begin`, `doRun`, `executeRun`

rwcarlsen replied:
Or maybe just `run` and `[run]Inner`, or `flexecute` and `flexecuteInner`.  Regardless, I was planning on renaming this function as well - but I figured its name will be informed by the rename for the class/system anyway.

fdkong replied:
I like `executeRun`

friedmud commented on framework/include/executioners/Runner.h:
```diff
@@ -0,0 +1,165 @@
+  /// code.
+  virtual Result gogogadget() = 0;
+
+  virtual void execute() override {}
```
Maybe make this `final`

friedmud commented on framework/doc/content/source/runners/Runner.md:
```diff
@@ -0,0 +1,184 @@
+  // When we record an inner/sub runner's result, we give it a label - which
+  // helps identify its placement/role within the runner heirarchy.
+  r.record("solve1", _inner_solve1->run());
+  r.record("solve2", _inner_solve2->run());
```
Since we have the outer wrapping `run()` function... can we just use that to do the recording instead?  This seems like it may be error-prone.

rwcarlsen replied:
Record is for building the result hierarchy.  And the parent/base class doesn't know anything about the child runners and details of particular runner implementations.  A runner calls its child/sub runners at its leisure - not at particular points we can know a prior.  As it goes along and results are generated they are accumulated into a result tree.  The only way to avoid this is to develop a shadow-hierarchy based on input file syntax that constructs the result tree to mirror runner structure - which seems more restrictive and a lot more complicated for what I would consider relatively small benefit.

friedmud replied:
Hmmm... I can see what you're saying... let me think about it a bit more.

friedmud commented on framework/src/base/MooseApp.C:
```diff
@@ -1163,6 +1174,25 @@ MooseApp::disableCheckUnusedFlag()
+void
+MooseApp::addRunner(std::shared_ptr<Runner> && runner)
+{
+  _runners[runner->name()] = runner;
```
Needs a check/error for name collisions

friedmud commented on framework/src/problems/FEProblemBase.C:
```diff
@@ -3693,6 +3693,53 @@ FEProblemBase::setCurrentExecuteOnFlag(const ExecFlagType & flag)
+    _displaced_problem->setCurrentlyComputingJacobian(false);
+
+  if (_control_warehouse.hasExecType(exec_type))
+    executeControls(exec_type);
```
`executeControls` again?

friedmud replied:
This function is 90% just the same as `execute()` below.  Are you sure this is needed?  Can we at least call `execute()` for most of it?

rwcarlsen replied:
The things that execute does actually cause errors when called from a runner context - because they assume some setup/stuff that doesn't occur outside the regular executioner stack.  I'm actually open to just making this function be empty - i.e. to be implemented in the future.  Because it's already quite incomplete in terms of what we would want for the new runner system.

friedmud replied:
Well - I think it's useful - even if incomplete.  Can you just add a boolean to `execute` to turn off the things you don't want to run right now?  Just trying to keep down duplication.

friedmud commented on test/include/executioners/SteadyWithPicardCheck.h:
```diff
@@ -26,11 +26,6 @@ class SteadyWithPicardCheck : public Steady
-   * Additional convergence check to be honored by PicardSolve
-   */
-  virtual bool augmentedFixedPointConvergenceCheck() const override;
-
```
What's the reason why this had to be removed?

rwcarlsen replied:
Nobody is using it.  I asked on the mailing list and scoured app repositories.  Griffin folks added it originally because they thought it would help with some stuff - but it didn't and now they don't use or need it.

friedmud replied:
👍

YaqiWang replied:
I removed it in Griffin for my proposed PR in MOOSE. We do not need it.

friedmud reviewed (dismissed):
Just a few comments to get started.

fdkong commented on framework/include/base/MooseApp.h:
```diff
@@ -66,6 +67,14 @@ class MooseApp : public ConsoleStreamInterface,
                  public libMesh::ParallelObject
 {
 public:
+  struct SolveConfig
```
Change `SolveConfig`  to `FixedPointSolveConfig` since it is unrelated to nonlinear solvers

fdkong commented on framework/src/actions/CreateRunnerAction.C:
```diff
@@ -0,0 +1,78 @@
+  // If enabled, automatically create a Preconditioner if the [Preconditioning] block is not found
+  if (_auto_preconditioning && !_awh.hasActions("add_preconditioning") &&
+      _moose_object_pars.isParamValid("solve_type"))
+    setupAutoPreconditioning();
```
I thought a runner could be a "flexible" executor that might not have `Preconditioning`

`Preconditioning` might be handled by "NonlinearSystem" ?

rwcarlsen replied:
This is because currently, the runner class is just inheriting from the executioner class - and so copies all the same setup.  This will change eventually.  It's just a matter of determining how and when that transformation occurs.

fdkong commented on unit/src/RunnerTests.C:
```diff
@@ -0,0 +1,39 @@
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "Runner.h"
```
Do we have a reference implementation of Runner? And build an input file for that as well?

rwcarlsen replied:
I do in some other branches, but those still have failing tests and issues that I'm trying to work out.  We wanted to get this up soon so that Yaqi can start building some of his stuff on it for some near term deadlines - while we work on reimplementing the fixedpoint/picard stuff in the new runner system in parallel.

fdkong reviewed (dismissed):
A reference implementation of Runner is following up?

friedmud commented:
I think that with some minor modifications we should be able to get this in.

rwcarlsen commented:
@friedmud - bass class/system name recommendation?  I think that is probably the biggest thing we need to finalize here.

friedmud commented:
Hard to say.

I really was serious about `Flexecutioner` - but that name doesn't seem to have caught fire.

It sounded like someone (@roystgnr ?) was suggesting `Executor`... it has a nice ring to it and could ultimately just replace `Executioner`.  Although I will be sad see that name go... maybe a clean break into a new name is a good idea (especially for clarity with our userbase).

If we go with `Executor` then we will probably want functions like `execute()`, `fulfill()`, `enact()`, `effect()`, `complete()`, `effectuate()`.

I don't really have a strong feeling about them.  `execute()` could be good as the non-virtual... and `fulfill()` might be good as the pure-virtual.

What do you think?

joshuahansel reviewed (commented):
I wasn't in the design discussions, so I'll defer that review to others.

rwcarlsen committed:
remove legacy params

rwcarlsen committed:
rename runner->executor

rwcarlsen committed:
rename gogogadget->run run->exec

rwcarlsen committed:
add executor action doc file

rwcarlsen committed:
make run function virtual

rwcarlsen committed:
fix vector allocation to occur during construction - this requires setting solve config before we parse input file and create objects

rwcarlsen committed:
undo bad search-replace

rwcarlsen committed:
add executor syntax doc file

moosebuild commented:
Job [Generate and verify coverage](https://civet.inl.gov/job/849663/) on 467b877 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>77ce47</th>
      <th colspan=4>#18966 467b87</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>81.64%</td>
      <td>81.56%</td>
      <td>-0.08%</td>
      <td>48.30%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>69808</td>
      <td>69853</td>
      <td>+45</td>
      <td>85</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15701</td>
      <td>15792</td>
      <td>+91</td>
      <td>91</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/18966/coverage_diff/framework.html)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/18966/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/18966/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/18966/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/18966/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/18966/coverage/contact)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/18966/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/18966/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/18966/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/18966/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/18966/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/18966/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/18966/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/18966/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/18966/coverage/navier_stokes)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/18966/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/18966/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/18966/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/18966/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/18966/coverage/rdg)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/18966/coverage/richards)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/18966/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/18966/coverage/tensor_mechanics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/18966/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 48.30% is less than the suggested 90.0%



This comment will be updated on new commits.

rwcarlsen committed:
fix more runner->executor search/replace mistakes

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/849947/) on 2a5d3b5 : invalidated by @rwcarlsen

rwcarlsen committed:
rename solve config to fixed point config

rwcarlsen committed:
make unused execute function final

rwcarlsen committed:
duplicate name check

rwcarlsen committed:
fix doc typo

rwcarlsen committed:
add test executor

friedmud committed:
Can read in all executor params and detect dependencies refs ##18180

friedmud committed:
Ordered creation of Executors is working refs #18180

friedmud committed:
Can now check for multiple executor roots refs #18180

friedmud committed:
Can now detect Executor cycles refs #18180

friedmud committed:
Add executor tests for everything refs #18180

friedmud committed:
Move Executors to their own directory refs #18180

friedmud committed:
Rework Executor base class and interfaces to be a bit more cosistent refs #18180

friedmud committed:
Fixup doco refs #18180

moosebuild commented:
Job [Mac Test](https://civet.inl.gov/job/870227/) on 317e11e : invalidated by @rwcarlsen

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/870233/) on 317e11e wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>7607a3</th>
      <th colspan=4>#18966 317e11</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>81.83%</td>
      <td>81.84%</td>
      <td>+0.01%</td>
      <td>85.39%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>71503</td>
      <td>71697</td>
      <td>+194</td>
      <td>228</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15875</td>
      <td>15912</td>
      <td>+37</td>
      <td>39</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/18966/coverage_diff/framework.html)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/18966/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/18966/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/18966/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/18966/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/18966/coverage/contact)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/18966/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/18966/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/18966/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/18966/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/18966/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/18966/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/18966/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/18966/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/18966/coverage/navier_stokes)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/18966/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/18966/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/18966/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/18966/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/18966/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/18966/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/18966/coverage/richards)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/18966/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/18966/coverage/tensor_mechanics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/18966/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 85.39% is less than the suggested 90.0%



This comment will be updated on new commits.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/874109/) on 649879a wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/18966/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/18966/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format ddf01c8a2c0df9a946ac6976b690b21d08aee95a`

rwcarlsen committed:
fix unity build

loganharbour commented on framework/src/base/Factory.C:
```diff
@@ -92,7 +92,7 @@ Factory::getADValidParams(const std::string & obj_name)
 Factory::create(const std::string & obj_name,
                 const std::string & name,
-                InputParameters & parameters,
+                const InputParameters & parameters,
```
🙌🏼

friedmud replied:
@loganharbour I bet you were surprised that I was _adding_ constness to MOOSE! ;-)

loganharbour replied:
I figured this was @rwcarlsen 😆

loganharbour reviewed (changes requested):
All very minor changes... let's get this in! Tried to find design suggestions, but no luck. Guess that's a good thing.

GiudGiud commented on framework/include/executors/Executor.h:
```diff
@@ -0,0 +1,166 @@
+    /// to recursively determine if there was any descendant executor failure.
+    bool converged = true;
+
+    /// Optional message detailing why a executor passed or failed (i.e. failed to converge).
```
```suggestion
    /// Optional message detailing why an executor passed or failed (i.e. failed to converge).
```

GiudGiud commented on framework/include/executors/Executor.h:
```diff
@@ -0,0 +1,166 @@
+    std::map<std::string, Result> subs;
+
+    /// Prints a full recursive output of this result object - including all
+    /// descendant's results.  If success_msg is true, then all result output
+    /// that contains a message will be printed even if it converged/passed.
```
```suggestion
    /// descendant's results.  If success_msg is true, then all results' output
    /// that contain a message will be printed even if it converged/passed.
```

rwcarlsen replied:
I think the original text is more clear.

GiudGiud commented on framework/include/executors/Executor.h:
```diff
@@ -0,0 +1,166 @@
+
+    /// Records results from sub/internal executors in a executor's result.  When
+    /// child-executors return a result object following their execution, this
+    /// function should be called to add that info into the result heirarchy.
```
```suggestion
    /// function should be called to add that info into the result hierarchy.
```

GiudGiud commented on framework/include/executors/Executor.h:
```diff
@@ -0,0 +1,166 @@
+
+    /// Returns false if any single executor in the current hierarchy of results
+    /// (i.e. including all child results accumulated recursively via record)
+    /// had a faild/unconverged return state.  Returns true otherwise.  This
```
```suggestion
    /// had a failed/unconverged return state.  Returns true otherwise.  This
```

GiudGiud commented on framework/include/executors/Executor.h:
```diff
@@ -0,0 +1,166 @@
+    return _result;
+  }
+
+  virtual bool lastSolveConverged() const override { return _result.convergedAll(); }
```
```suggestion
  /// Whether the executor and all its sub-executors passed / converged
  virtual bool lastSolveConverged() const override { return _result.convergedAll(); }
```

GiudGiud commented on framework/include/executors/Executor.h:
```diff
@@ -0,0 +1,166 @@
+             ((success_msg || !converged) && !reason.empty() ? ": " + reason : "");
+    }
+    std::string _name;
+  };
```
is there a way to mark that this is the end of Result ?
I had a double take when I saw the constructor below

rwcarlsen replied:
We don't have any conventions.  I'm inclined to leave it.  Just one of the joys of C++ syntax - nested namespace/class scopes.

GiudGiud commented on framework/src/problems/FEProblemBase.C:
```diff
@@ -3693,6 +3693,53 @@ FEProblemBase::setCurrentExecuteOnFlag(const ExecFlagType & flag)
+    _displaced_problem->setCurrentlyComputingJacobian(false);
+
+  if (_control_warehouse.hasExecType(exec_type))
+    executeControls(exec_type);
```
execControls has to be executed before UOs
see this PR by @andrsd 
https://github.com/idaholab/moose/pull/18508

GiudGiud commented on framework/include/executors/ExecutorInterface.h:
```diff
@@ -0,0 +1,57 @@
+   * Get an Executor based on a parameter name
+   *
+   * @param param_name The name of the parameter
+   * @param index Only used for vectors of executor names
```
```suggestion
```

GiudGiud commented on framework/include/executors/ExecutorInterface.h:
```diff
@@ -0,0 +1,57 @@
+   * Get an Executor based on its actual name
+   *
+   * @param executor_name the actual name of the Executor
+   * @param index Only used for vectors of executor names
```
```suggestion
```

GiudGiud commented on framework/include/executors/ExecutorInterface.h:
```diff
@@ -0,0 +1,57 @@
+  Executor & getExecutorByName(const ExecutorName & executor_name) const;
+
+private:
+  const MooseObject * _ei_moose_object;
+
+  MooseApp & _ei_app;
```
docstrings

loganharbour replied:
did you even look at my review when you reviewed?

rwcarlsen replied:
I'm not sure what a docstring could possibly add here that isn't trivially apparent via the C++ type, moose conventions, and the code.

GiudGiud commented on framework/src/actions/AddExecutorAction.C:
```diff
@@ -0,0 +1,32 @@
+
+registerMooseAction("MooseApp", AddExecutorAction, "add_executor");
+
+defineLegacyParams(AddExecutorAction);
```
```suggestion
```

GiudGiud reviewed (commented):
Add doc page for the add Executor action and ReadExecutorParamsAction too

GiudGiud commented on framework/include/executors/Executor.h:
```diff
@@ -0,0 +1,166 @@
+/// the same pattern as with mesh generators.  Subclass here and implement your
+/// own run function if you need special or fancy functionality that
+/// isn't supported by the default executors available.
+class Executor : public Executioner, public ExecutorInterface
```
```suggestion
class Executor : public Executioner, public ExecutorInterface
```
can we avoid inheriting the Executioner?
Stuff like calls to executor.feProblem should really not be possible

GiudGiud replied:
well it makes sense for this particular one, since it has to do the solve. Does it make sense for a Refine one?

friedmud replied:
I believe that this is just a transient situation.  It was the easiest way to "rig" Executors into the current system.

rwcarlsen replied:
Yep.  They could be separated.  Since eventually executioners will be killed, it didn't seem worth the trouble for now.

friedmud committed:
Update framework/src/actions/AddExecutorAction.C

Co-authored-by: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

friedmud committed:
Update framework/include/executors/ExecutorInterface.h

Co-authored-by: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

friedmud committed:
Update framework/include/executors/ExecutorInterface.h

Co-authored-by: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

friedmud committed:
Update framework/include/base/MooseApp.h

Co-authored-by: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

friedmud commented on framework/include/base/MooseApp.h:
```diff
@@ -35,6 +35,8 @@
 // Forward declarations
 class Executioner;
+class Executor;
+class NullExecutor;
```
For optional coupling.

Or are you asking why we need the forward declaration?  Because it's referenced down below.

friedmud committed:
Update framework/include/actions/ReadExecutorParamsAction.h

Co-authored-by: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

friedmud committed:
Update framework/include/actions/AddExecutorAction.h

Co-authored-by: Logan Harbour <loganhharbour@gmail.com>

friedmud committed:
Update framework/include/actions/ReadExecutorParamsAction.h

Co-authored-by: Logan Harbour <loganhharbour@gmail.com>

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/884315/) on 5358d6a wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/18966/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/18966/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 9445acccf9900eba978b24465b8886e32fa8bf4b`

fdkong commented on framework/include/base/MooseApp.h:
```diff
@@ -939,6 +1002,24 @@ class MooseApp : public ConsoleStreamInterface,
+  std::shared_ptr<Executor> _executor;
+
+  /// Pointers to all of the Executors for this run
+  std::map<std::string, std::shared_ptr<Executor>> _executors;
```
Are there next level executors?

friedmud replied:
I'm not sure I understand?

fdkong replied:
I meant: whether or not `std::map<std::string, std::shared_ptr<Executor>> _executors` are children of `std::shared_ptr<Executor> _executor`?

rwcarlsen replied:
You can code executors to store subs/children however you want.  The system doesn't presuppose any structure.

rwcarlsen commented on framework/include/executioners/SecantSolve.h:
```diff
@@ -30,6 +30,9 @@ class SecantSolve : public FixedPointSolve
   virtual void allocateStorage(const bool primary) override final;
 
 private:
+  bool _need_pp_values_init = true;
```
Yeah - this got moved out for reasons that are now moot - so I'll remove the separate init and put it back in the constructor.

rwcarlsen commented on framework/include/executioners/SecantSolve.h:
```diff
@@ -30,6 +30,9 @@ class SecantSolve : public FixedPointSolve
 
 private:
+  bool _need_pp_values_init = true;
+  void initPostprocessorValues();
```
Just removed it altogether

fdkong commented on framework/src/base/MooseApp.C:
```diff
@@ -1163,6 +1178,162 @@ MooseApp::disableCheckUnusedFlag()
+}
+
+void
+MooseApp::recursivelyCreateExecutors(const std::string & current_executor_name,
```
I would simply have a parent executor to create its own children. The job of MooseApp is add the highest level executor. 

This idea will allow a customized executor to define its fashion to create its children.   Another advantage of this idea is that we then allow a delayed creation of children, which might be useful for some applications.

rwcarlsen replied:
This is just for creating the objects.  They do "fetch" their own children within themselves.  They plumb themselves together - this is mostly just for creating them in the right order so we can bind them together in their constructors - we need to handle deps so we know which ones to create first, etc.

friedmud replied:
One Executor can be the child of multiple others.

fdkong replied:
>This is just for creating the objects.

I understood. How about I do not want to create my children executors when the outtest extupor is created?  

I could decide if I want to create an inner solver or not based on the result from the previous solver.  For example, if the init solver alway fails, why I need to waste memory to create the following-up solvers?

fdkong commented on test/tests/executors/multiple/test.i:
```diff
@@ -0,0 +1,55 @@
+  [my_first_executor]
+    type = BinaryTestExecutor
+    inner1 = i1
+    inner2 = i2
```
Could we do ` inner = 'i1 i2'`?

rwcarlsen replied:
If you write an executor to do that sure.

friedmud replied:
This is just a test object... but we didn't add "vector" capability for this system yet.

fdkong reviewed (dismissed):
IMHO, it is not the job of the App to create all executors. Instead, the parent executor should create its own children when needed. This allow flexibility for a customized  executor, where it can determine when and what it wants to create

rwcarlsen committed:
doc comments and cleanups

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/884627/) on 6588656 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/18966/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/18966/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 9445acccf9900eba978b24465b8886e32fa8bf4b`

rwcarlsen commented on framework/doc/content/source/executors/Executor.md:
```diff
@@ -0,0 +1,185 @@
+  []
+  [init]
+    type = Init
+    inner = refine
```
executors are responsible for having their own arbitrary internal/sub executor structure.  There is nothing special about "inner" here. - It's just how it was coded.  You could write an executioner that takes vectors of names like `peon1=foo` or `underlings = 'foo bar baz'`.  It's not baked into the system how you structure sub executors.  And we create them following the moose convention where an input file block equals an object created by the factory.

friedmud replied:
Yes - true tree

friedmud replied:
AND - each Executor can be in the tree multiple times and in multiple places.

rwcarlsen commented on framework/include/base/MooseApp.h:
```diff
@@ -66,6 +68,14 @@ class MooseApp : public ConsoleStreamInterface,
                  public libMesh::ParallelObject
 {
 public:
+  struct FixedPointConfig
```
This is just an artifact of needing to communicate multiapp input file config to the fixed point solve.  It may go away at some point.  But it is here for now and is stored and accessed via mooseapp - so may as well be defined there.

rwcarlsen committed:
clang format one of derek's old commits

rwcarlsen committed:
make executeAllObjects just be an empty stub

rwcarlsen committed:
Update framework/include/executors/Executor.h

Co-authored-by: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

rwcarlsen committed:
typo

Co-authored-by: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

rwcarlsen committed:
typo

Co-authored-by: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

rwcarlsen committed:
doc comment

Co-authored-by: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

rwcarlsen committed:
fixup! make executeAllObjects just be an empty stub

rwcarlsen committed:
add missing docs

moosebuild commented:
Job [Mac Test](https://civet.inl.gov/job/887268/) on 55e6286 : invalidated by @rwcarlsen

YaqiWang commented on framework/src/executors/Executor.C:
```diff
@@ -0,0 +1,54 @@
+    _begin_flag(getParam<ExecFlagType>("begin_exec_flag")),
+    _end_flag(getParam<ExecFlagType>("end_exec_flag"))
+{
+  if (!parameters.isParamSetByUser("begin_exec_flag"))
```
Should here be `if (parameters.isParamSetByUser("begin_exec_flag"))`?

rwcarlsen replied:
Nope.  We set it to an auto-generated special name only if the user didn't set it.
